subsys/net/ip/tcp2.c

```diff
@@ -832,6 +954,89 @@ static struct tcp *tcp_conn_new(struct net_pkt *pkt)
 	return conn;
 }
 
+static void trim_unacked_data(struct tcp *conn)
+{
+	/* Make sure there is no empty buffers in the beginning
+	 * of the chain as we just pulled data from it (so there
+	 * should not be any)
+	 */
+	if (conn->unacked_data->buffer &&
+	    conn->unacked_data->buffer->len == 0) {
+		/* TODO: Fix net_pkt_pull() so that this check can be removed.
+		 */
+		struct net_buf *buf = conn->unacked_data->buffer;
+
+		while (buf) {
+			if (buf->len) {
+				break;
+			}
+
+			conn->unacked_data->buffer = buf->frags;
+			buf->frags = NULL;
+
+			net_buf_unref(buf);
+			if (!conn->unacked_data->buffer) {
+				break;
+			}
+
+			buf = conn->unacked_data->buffer->frags;
+		}
+	}
+}
+
+static u32_t calc_acked_data(struct tcp *conn, struct tcphdr *th)
+{
+	u32_t acked_len = th_ack(th) - conn->last_seq;
+
+	if (!conn->unacked_data || acked_len == 0) {
+		return 0U;
+	}
+
+	NET_DBG("Peer acked %u bytes", acked_len);
+
+	return acked_len;
+}
+
+static void update_acked_data(struct tcp *conn, u32_t acked_len)
+{
+	int ret;
+
+	/* If there is any pending data, remove them from re-send as peer
+	 * just acked it.
+	 */
+	net_pkt_cursor_init(conn->unacked_data);
+	net_pkt_set_overwrite(conn->unacked_data, true);
+
+	/* TODO: Check that pull would succeed beforehand. For this we need
+	 * to know the length of unacked_data. But the net_pkt_get_len() is
+	 * expensive, and as we need to know the length information in many
+	 * places, fix this by keeping track of length separately. The length
+	 * calculations in tcp_send_queued_data() would need to be changed too.
+	 */
+	ret = net_pkt_pull(conn->unacked_data, acked_len);
```

Comment:
Shall we move it into tcp_pkt_trim()? Actually tcp_pkt_pull() would be a better name for tcp_pkt_trim().